Drive level 2 remake

AryanKammili opened the pull request.

AryanKammili committed:
added swerve code

AryanKammili committed:
Fixed naming convention

AryanKammili committed:
IO layer fully implemented

AryanKammili committed:
removed unsused comments/ added robocontainer

AryanKammili committed:
added PP constants/tested auton logic

AryanKammili committed:
sim testing

mduffor commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+    // TODO: Needs to be tuned //
+    public static final double kDriftRate = 1;
+
+    public static final SwerveModuleHardwareConfig kFrontLeft = new SwerveModuleHardwareConfig(
```
Need a comment on how to get the rotation values to put in the code for each module.

AryanKammili replied:
already in slides

mduffor commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+
+    public static final boolean kInvertAzimuths = true;
+
+    // Make sure to change this gains after each time of tuning //
```
Where does the magic number 660.0 come from?

mduffor commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+        PIDController azimuthController,
+        SimpleMotorFeedforward azimuthFF) {}
+
+    public static final ModuleControlConfig kModuleControllerConfigs = RobotBase.isReal() ? 
```
Need information on how these tuning values are calculated.  In three years, when a rookie is looking at this code, how will they figure out how to utilize it?

AryanKammili replied:
Its the values from PID tuning but sure I can add comments

mduffor commented on src/main/java/frc/robot/drive/ModuleIOSim.java:
```diff
@@ -0,0 +1,95 @@
+public class ModuleIOSim implements ModuleIO {
+
+    private DCMotorSim driveMotor = 
+    new DCMotorSim(
```
Where do the magic numbers 0.004 and 0.025 (J moment of inertia) come from?

mduffor commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+    public static final double kAzimuthGearRatio = 150.0 / 7.0;
+    public static final double kDriveGearRatio = 6.75 / 1.0;
+    public static final double kWheelRadiusMeters = 5.08 / 100.0;
+    public static final double kCircumfrenceMeters = 2 * Math.PI * kWheelRadiusMeters;
```
kCircumfrenceMeters should probably be named kWheelCircumferenceMeters, so it can be better understood in context in the rest of the code (and have circumference spelled correctly ;-) )

AryanKammili replied:
done!

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.utils.debugging.LoggedTunableNumber;
+
+public class Module extends SubsystemBase {
```
Why do we have Module extend from SubsystemBase?  Is there any functionality we are using there?

AryanKammili replied:
It was earlier because of the way it was set up earlier. It can be removed now

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * Reset the posistion of the azimuth encoder
```
spelling

mduffor replied:
errr.... you did a mass rename to "posistion".  I hope that's the reason at least.

AryanKammili replied:
yeah i cant speel position

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+     * @return set state
+     */
+    public SwerveModuleState setSwerveStatewithAccel(SwerveModuleState state, double accel){
+        setDriveAcceleration(accelerationSetpointMPSS);
```
The 'accel' parameter isn't used.  Should that be what is passed to setDriveAcceleration()?

AryanKammili replied:
sorry was testing something but is now removed

mduffor commented on src/main/java/frc/robot/drive/ModuleIO.java:
```diff
@@ -0,0 +1,43 @@
+    @AutoLog
+    public static class SwerveModuleInputs {
+        public boolean driveConnected = true;
+        public double drivePosistionM = 0.0;
```
What purpose does drivePosistionM actually serve?

AryanKammili committed:
Tuning chnages

mduffor commented on src/main/java/frc/robot/drive/ModuleIOKraken.java:
```diff
@@ -0,0 +1,213 @@
+    azimuthVoltageControl = new VoltageOut(0);
+    driveVoltageControl = new VoltageOut(0);
+
+    // Closed - Loop control //
```
Need comments here on how you are creating these controllers, and what parameters you are passing in.

AryanKammili replied:
sure

mduffor commented on src/main/java/frc/robot/drive/ModuleIOKraken.java:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  @Override
+  public void setDriveVelocity(double velocityMPS, double feedforward) {
```
feedforward param is ignored here.  Should it be?

AryanKammili replied:
again due to testing and is fixed

mduffor commented on src/main/java/frc/robot/drive/ModuleIOKraken.java:
```diff
@@ -0,0 +1,213 @@
+    TalonFXConfiguration driveConfig = new TalonFXConfiguration();
+    drive.getConfigurator().apply(driveConfig);
+    driveConfig.CurrentLimits.StatorCurrentLimitEnable = true;
+    driveConfig.CurrentLimits.StatorCurrentLimit = 80;
```
Can we link to documentation where we define what stator and supply current limits we use at Prosper?

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+
+    public static final LoggedTunableNumber azimuthP = new LoggedTunableNumber("Module/Azimuth/kP", DriveConstants.kModuleControllerConfigs.azimuthController().getP());
+    public static final LoggedTunableNumber azimuthD = new LoggedTunableNumber("Module/Azimuth/kD", DriveConstants.kModuleControllerConfigs.azimuthController().getD());
+    public static final LoggedTunableNumber azimuthS = new LoggedTunableNumber("Module/Azimuth/kS", DriveConstants.kModuleControllerConfigs.azimuthFF().getKa());
```
This should be getKs(), not getKa(), correct?

AryanKammili replied:
typo corrected before deployed to robot

mduffor commented on src/main/java/frc/robot/drive/ModuleIO.java:
```diff
@@ -0,0 +1,43 @@
+
+    public default void setAzimuthPID(double kP, double kI, double kD) {}
+
+    public default void setAzimuthFF(double kS, double kV, double kA) {}
```
Need a comment indicating that setAzimuthPID's kI, and setAzimuthFF kV and kA are not utilized by the code.  Add similar comments for the Drive calls indicating which parameters are ignored.

mduffor commented on src/main/java/frc/robot/drive/ModuleIOKraken.java:
```diff
@@ -0,0 +1,213 @@
+      driveStatorCurrent,
+      driveTempC).isOK();
+
+    inputs.drivePosistionM = drivePosistion.getValueAsDouble();
```
drivePosistionM isn't used anywhere, is it?

mduffor commented on src/main/java/frc/robot/drive/ModuleIOKraken.java:
```diff
@@ -0,0 +1,213 @@
+    azimuth.setPosition(
+      Rotation2d.fromRotations(
+      absolutePosistionSignal.getValueAsDouble()).minus(angleOffset).getRotations(), 
+      2.5);
```
The 2.5, assuming it is "timeout in seconds" needs to be assigned to a named variable/constant so the code can be better understood.

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+
+    private String nameKey;
+
+    public Module(String key, ModuleIO config){
```
"instance" would be a better variable name here than "config"

AryanKammili replied:
okay

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+
+        LoggedTunableNumber.ifChanged(
+            hashCode(), () -> {
+                driveFeedforward = new SimpleMotorFeedforward(driveS.get(), driveV.get(), driveA.get());
```
Why is a driveFeedforward class used here and the driveS isn't passed down to the ModuleIO such that ModuleIOKraken can then apply it to Slot0Configs kS?  Are you double-applying feed forward since PositionVoltage already handles it?  Or is this class's results simply ignored because the setDriveVelocity's feedforward param is ignored?

AryanKammili replied:
I was testing something and forgot to comment it out, sorry for the confusion

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+     */
+    public void runLinearCharacterization(double inputVolts) {
+        setAzimuthPosistion(Rotation2d.fromRotations(0));
+        setDriveVelocity(null);
```
You pass a null to setDriveVelocity to switch from velocity mode to volts mode.  A boolean defining the mode would be clearer.  But for safety, shouldn't you just call `setDriveVelocity(null);` within setDriveVolts() just to make sure it always gets forced off?

AryanKammili replied:
Its been working as is as I ran multiple system identification tests and had no problem. After all setting drive velocity to null means that in periodic its just ignored and doesn't do anything

mduffor commented on src/main/java/frc/robot/drive/Module.java:
```diff
@@ -0,0 +1,217 @@
+    /**
+     * Stop swerve module
+     */
+    public void stop(){
```
Do you also need to set your velocities and accelerations to null?

mduffor commented on src/main/java/frc/robot/drive/Drive.java:
```diff
@@ -0,0 +1,409 @@
+    }
+
+    public boolean isSpeedOptimized(SwerveModuleState state, SwerveModuleState optimizedState) {
+        return state.speedMetersPerSecond != optimizedState.speedMetersPerSecond;
```
Don't directly compare floats.  Use an epsilon comparison within a reasonable epsilon.

mduffor commented on src/main/java/frc/robot/drive/Drive.java:
```diff
@@ -0,0 +1,409 @@
+        setSwerve(new ChassisSpeeds());
+    } 
+
+    public void resetGyro() {
```
You need a comment for how the robot needs to be oriented when this is called.

mduffor commented on src/main/java/frc/robot/drive/Drive.java:
```diff
@@ -0,0 +1,409 @@
+
+        field.setRobotPose(getEstimatedPose());
+
+        ChassisSpeeds teleopSpeeds = teleopController.computeChassisSpeeds((getEstimatedPose().getRotation()), getChassisSpeeds());
```
It would help to have a comment here that describes how the teleopController integrates with the Drive subsystem.

mduffor commented on src/main/java/frc/robot/drive/controllers/TeleopController.java:
```diff
@@ -0,0 +1,89 @@
+package frc.robot.drive.controllers;
```
This class could use some documentation/comments to explain it to people new to the code.

mduffor commented on src/main/java/frc/robot/drive/Drive.java:
```diff
@@ -0,0 +1,409 @@
+
+        SmartDashboard.putData(field);
+
+        headingController.setHeadingGoal(() -> headingGoal);
```
Need an explanation on how to use heading goal, and what it overrides.  It doesn't appear to be completely hooked up at the moment.

AryanKammili replied:
Yeah hooked it up

mduffor commented on src/main/java/frc/robot/drive/controllers/HeadingController.java:
```diff
@@ -0,0 +1,100 @@
+  
+    private Supplier<Rotation2d> goal;
+    // Me: Sign that there may be something fundamentally wrong with the code :)
+    // Me at a later date: Yeah there is a fundamental mistake with the code :)
```
Is there a source reference for what inspired this code?

AryanKammili replied:
6328

mduffor commented on src/main/java/frc/robot/drive/controllers/HeadingController.java:
```diff
@@ -0,0 +1,100 @@
+    private double invert;
+
+    public HeadingController() {
+        invert = (RobotBase.isReal()) ? -1.0 : 1.0;
```
Why is this inverting based on robot mode?

mduffor commented on src/main/java/frc/robot/drive/controllers/HeadingController.java:
```diff
@@ -0,0 +1,100 @@
+
+    public double getControlledHeadingOutput(Rotation2d robotRotation) {
+        return (snapController.getGoal().equals(snapController.getSetpoint())
+            && snapController.getPositionError() < 4.5)
```
4.5 is a magic number.  Need to move it to a descriptively named variable.

mduffor commented on src/main/java/frc/robot/drive/controllers/HeadingController.java:
```diff
@@ -0,0 +1,100 @@
+        double goalErrorDegrees = snapController.getGoal().position - robotRotation.getDegrees();
+        Logger.recordOutput("Drive/setpointErrorDegrees", setpointErrorDegrees);
+        Logger.recordOutput("Drive/goalErrorDegrees", goalErrorDegrees);
+        if(Math.abs(goalErrorDegrees) < 0.75) output *= 0.0;
```
Another magic number that needs to be named and in a constants file.

AryanKammili replied:
Fixed heading controller

AryanKammili committed:
quality of life changes

AryanKammili committed:
fixes to get code to wokr, very messy needs clean up

Manyata-Patel committed:
MP - Updated WPILib to 2025.2.1

AryanKammili commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -33,12 +45,30 @@ public RobotContainer() {
             default:
                 // Instantiate subsystems that are driven by playback of recorded sessions. (IO modules)
+                drive = new Drive( new Module[] {
+                    new Module("FL", new ModuleIO(){}),
```
Needs to be tested with replay

AryanKammili commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -33,12 +45,30 @@ public RobotContainer() {
+                    new Module("FR", new ModuleIOSim()),
+                    new Module("BL", new ModuleIOSim()),
+                    new Module("BR", new ModuleIOSim())
+                }, new GyroHardware());
```
Added Gyro Sim

AryanKammili committed:
Sim closer to working, need rotation fully working. Clean of code workign through duffys commnets

AryanKammili committed:
Merge branch 'drive-level-2-remake' of https://github.com/Prosper-FRC/template-talon_standard into drive-level-2-remake

AryanKammili commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -60,12 +90,17 @@ public RobotContainer() {
 
 
         // Pass subsystems to classes that need them for configuration
+        drive.acceptJoystickInputs(
```
The robots front is considered "x" while the left and right is considered the "y".

AryanKammili replied:
anshuls code is set up the same way

AryanKammili commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -81,12 +116,21 @@ public Command getAutonomousCommand() {
     }
 
     private void configureStateTriggers() {
+        new Trigger(DriverStation::isEnabled).onTrue(Commands.runOnce(() -> drive.resetAllEncoders()));
```
probably gonna remove it, it shouldn't be necessary but I was using it for testing earlier on

AryanKammili commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -81,12 +116,21 @@ public Command getAutonomousCommand() {
+    driverController.x().onTrue(Commands.runOnce(() -> {drive.resetGyro();}));
 
-    }
+    driverController.povUp().onTrue(drive.setDriveStateCommandContinued(DriveState.SNIPER_UP)).onFalse(drive.setDriveStateCommand(DriveState.TELEOP));
```
done

AryanKammili commented on src/main/java/frc/robot/drive/GyroHardware.java:
```diff
@@ -0,0 +1,43 @@
+    @AutoLog
+    public static class GyroInputs {
+        public boolean gyroConnected = true;
+        public Rotation2d yaw = new Rotation2d();
```
fixed

AryanKammili commented on src/main/java/frc/robot/drive/GyroHardware.java:
```diff
@@ -0,0 +1,43 @@
+
+    public void updateInputs(GyroInputs inputs){
+        inputs.gyroConnected = BaseStatusSignal.refreshAll(yaw).isOK();
+        inputs.yaw = Rotation2d.fromDegrees(yaw.getValueAsDouble());
```
fixed

AryanKammili replied:
Documentation states that getYaw() gets the value in degrees

AryanKammili commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+
+    public static final int kGyroPort = 10;
+
+    public static final double kMaxLinearSpped = 4.8;
```
okay

AryanKammili commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+
+    public static final double kMaxLinearSpped = 4.8;
+    public static final double kMaxLinearAcceleration = 9.6;
+    public static final double kMaxRotationalSpeed = kMaxLinearSpped * kDrivebaseRadius;
```
sorry kept it in case but thought this was correct because I asked my physics teacher. Guess he's wrong lol

AryanKammili commented on src/main/java/frc/robot/drive/DriveConstants.java:
```diff
@@ -0,0 +1,91 @@
+    public static final double kMaxRotationalSpeed = kMaxLinearSpped * kDrivebaseRadius;
+    public static final double kMaxRadiansPS = Math.toRadians(1320);
+
+    public static final double kAzimuthGearRatio = 150.0 / 7.0;
```
It will be the same this year

AryanKammili committed:
Fixing up to template standards, almost done

Fixed Heading controller to work better and is a little bit easier to read, teleop controlelr added controller, fixed most naming issues, updated alliance flip util, adding comments to explain code

AryanKammili commented on src/main/java/frc/robot/drive/Drive.java:
```diff
@@ -0,0 +1,409 @@
+     * @param state state to pass in
+     * @return command
+     */
+    public Command setDriveStateCommandContinued(DriveState state){
```
It is used in robot container for SNIPER mode

AryanKammili commented on src/main/java/frc/robot/drive/Drive.java:
```diff
@@ -0,0 +1,409 @@
+            case STOP:
+                desiredSpeeds = null;
+                for(Module module : modules) {
+                    module.setDriveVolts(0.0);
```
agreed

AryanKammili committed:
fixing things that duffy wanted (comments)

AryanKammili committed:
SWERVE DRIFT IS DEAD IN SIM

AryanKammili committed:
Constants change for 9105 drivebase

AryanKammili committed:
Merge branch 'drive-level-2-remake' of https://github.com/Prosper-FRC/template-talon_standard into drive-level-2-remake

mduffor commented on src/main/java/frc/robot/drive/controllers/TeleopController.java:
```diff
@@ -0,0 +1,105 @@
+    public ChassisSpeeds computeChassisSpeeds(Rotation2d robotAngle, ChassisSpeeds currentRelativeSpeeds){
+
+        // Scales the filtered value if the person driving wanted a slower or faster robot //
+        double xChangedDemand = linearScalar.get() * applyDeadband(xSupplier.getAsDouble(), linearDeadband.get());
```
Why is the linear scalar applied before exponent?

mduffor commented on src/main/java/frc/robot/drive/controllers/TeleopController.java:
```diff
@@ -0,0 +1,105 @@
+        double omegaChangedDemand = omegaScalar.get() * applyDeadband(omegaSupplier.getAsDouble(), omegaDeadband.get());
+
+        // Makes sure that the exponenet is a whole number //
+        int linearExp = (int) Math.round(linearInputsExponent.get());
```
Why are exponents rounded to whole numbers?

mduffor commented on src/main/java/frc/robot/drive/controllers/TeleopController.java:
```diff
@@ -0,0 +1,105 @@
+        int rotationExp = (int) Math.round(rotationInputsExponent.get());
+
+        // Converts the value to m/s for the chassis speeds //
+        double xVelocityMPS = DriveConstants.kMaxLinearSpeed * Math.pow(xChangedDemand, linearExp);
```
Why are the linear and exponent calculations not clamped between 0 and 1 before applying to velocity?  If the values are set incorrectly, you could request a faster-than-max speed from the drive system.  I would suggest doing all of the control adjustments and storing them into variables, and only then after everything has been adjusted do you multiply by kMaxLinearSpeed.

AryanKammili committed:
working offsets -> need to re-ID some motors

AryanKammili committed:
RON PULL THIS CODE BOI

Manyata-Patel committed:
MP - Updated auton code so it works (actually changes state to auton in code)

Manyata-Patel committed:
Merge branch 'drive-level-2-remake' of https://github.com/Prosper-FRC/template-talon_standard into drive-level-2-remake

AryanKammili committed:
Drivebase updaed Jack!

barbute committed:
Merge branch 'drive-level-2-remake' into drive-level-2-remake

barbute committed:
Merge pull request #19 from Prosper-FRC/drive-level-2-remake

Merge 9105 swerve code back to templates

AryanKammili committed:
Vision code added to drive L2
